### ISSUE TITLE
Can I generate imports without the import alias (or without the long underscored import path)?

### ISSUE THREAD

**Issue by meling** [gogo/protobuf#338]
In our plugin, we have until now used a custom `GenerateImports(file *generator.FileDescriptor)` method, but it turned out to be unreliable in certain scenarios (see (*) below). So we wanted to switch to using the `generator.NewPluginImports(g.Generator)` thing. However, right now, I'm getting 
```go
import golang_org_x_net_trace "golang.org/x/net/trace"
import google_golang_org_grpc_codes "google.golang.org/grpc/codes"
import google_golang_org_grpc_status "google.golang.org/grpc/status"
```
which doesn't work for us since our plugin is generated from template code, which can't use the `Single.Use()` method. It would require us to hardcode these underscored import paths into our templates, which would be quite tedious, error prone, and make our templates quite unreadable.

My question is: is there a way to generate these imports without the underscored import alias??

(PS: I don't quite understand the need for this import alias thing, but I'm sure there is an explanation. But is it always necessary?)

(*) There are some more details concerning this [issue here](https://github.com/relab/gorums/issues/38), but hopefully, the above description should be clear. Here is an example [template](https://github.com/relab/gorums/blob/master/dev/calltype_common_definitions.tmpl) being used by our plugin.

**Comment by awalterschulze** [comment]
As far as I remember the long alias names in gogoprotobuf's generated code is simply a way to avoid most of the package name collisions and try to guarantee some uniqueness. You are welcome to try and change the import logic in gogoprotobuf, but its not fun and I would want a few production projects to test the change, since its quite convoluted.

Here are a couple of work arounds:

1) You can build Single.Use() into the template object as separate methods:

  - ImportContext
  - ImportCodecs

2) Or you could add a function or method to the object that takes a string:

  - {{import "context"}}.Context
  - {{import  "golang.org/x/net/trace"}}

3) Parse your generated go code for imports and add them to the ast, before outputting the generated code.

Sorry.  I know this is not the answer you were hoping for.

**Comment by meling** [comment]
I've tried to make a few small changes to the import logic that solves my problem. I've uploaded a PR with my fix, but clearly it would require much more scrutiny.

**Comment by meling** [comment]
Just found an interesting project related to NATS that is also generating code based on templates; see [here](https://github.com/chop-dbhi/nats-rpc). The trick they do to avoid the issues with the import statements they do is to simply generate the output file separately from the `pb.go` file generated by `protoc`. Not sure that would work for us though, since we rely on a different options, that wouldn't be found by the standard `protoc` compiler. But it might be worth checking it out some, to see if we can learn some tricks from their generator.

**Comment by meling** [comment]
Closing as the issue has been resolved with #341.

**Comment by awalterschulze** [comment]
Awesome work. Thanks again.

**Comment by awalterschulze** [comment]
On the `dev` branch golang/protobuf made a huge change.  When I merge this back in, then we start getting long import names again.  I played around a little, but I want to rather focus on getting us back up to date.  If someone wants to pick this up again, I would be really grateful.  Thanks
https://github.com/gogo/protobuf/commit/aa7e6f4b58a59e25ca6d89e2a971fe199df9aa8a
https://github.com/golang/protobuf/commit/9d4962b4dc40a899c435fe1aaec48e683b4300ef

**Comment by awalterschulze** [comment]
It seems more import changes are `dev` branch that I have not had time to merge in again.  Lets wait to merge those in before someone tries to fix this again.
For example:
https://github.com/golang/protobuf/commit/3b4abe1a0672c5916c1937b8817dde8aeb579fe5

**Comment by awalterschulze** [comment]
The `dev` branch is now up to date and the problem persists.
@meling if you would like to look at this again, I would really appreciate it.
Maybe this is something we should try to upstream to golang/protobuf so that this does not happen again?

**Comment by awalterschulze** [comment]
I agree. In goderive I found doing imports much simpler just doing as you
suggested.

It would be great if you could bring this up with them as well.

**Comment by chideat** [comment]
With the latest gogo, there still have problems.
I got an proto package named `error` then the generated code is 
```go
import error "git.bdmd.com/server/his/doctorstation/pkg/types/rxthinking/his/v1/error"
```
which conflict with the buildin `error` type. Some work should be do with buildin types.

**Comment by dsnet** [comment]
In the upstream `golang/protobuf` generator, these set of PRs addressed this problem:
* https://github.com/golang/protobuf/commit/31e0d063dd98c052257e5b69eeb006818133f45c#diff-67bd54535d8241acbdd7a7d3fd9005d8
* https://github.com/golang/protobuf/commit/ddf22928ea3c56eb4292a0adbbf5001b1e8e7d0d#diff-67bd54535d8241acbdd7a7d3fd9005d8

I don't know if they've been cherry-picked into `gogo/protobuf`'s generator.
